SU2_CFD/include/numerics_simd/CNumericsSIMD.hpp

```diff
@@ -0,0 +1,711 @@
+
+/*--- Static matrix types. ---*/
+template<class Type, size_t Rows, size_t Cols>
+using Matrix = C2DContainer<unsigned long, Type, StorageType::RowMajor, Type::Align, Rows, Cols>;
+
+template<size_t Rows, size_t Cols = Rows> using MatrixInt = Matrix<Int, Rows, Cols>;
+template<size_t Rows, size_t Cols = Rows> using MatrixDbl = Matrix<Double, Rows, Cols>;
+
+/*!
+ * \class CNumericsSIMD
+ * \brief Base class to define the interface.
+ */
+class CNumericsSIMD {
+public:
+  /*!
+   * \brief Interface for edge flux computation.
+   * \param[in] iEdge - The starting edge for flux computation.
+   * \param[in] config - Problem definitions.
+   * \param[in] geometry - Problem geometry.
+   * \param[in] solution - Solution variables.
+   * \param[in] updateType - Type of update done on vector and matrix.
+   * \param[in,out] vector - Target for the fluxes.
+   * \param[in,out] matrix - Target for the flux Jacobians.
+   */
+  virtual void ComputeFlux(unsigned long iEdge,
+                           const CConfig& config,
+                           const CGeometry& geometry,
+                           const CVariable& solution,
+                           UpdateType updateType,
+                           CSysVector<su2double>& vector,
+                           CSysMatrix<su2mixedfloat>& matrix) const = 0;
+
+  /*! \brief Destructor of the class. */
+  virtual ~CNumericsSIMD(void) = default;
+};
+
+template<size_t nDim>
+FORCEINLINE Double dot(const VectorDbl<nDim>& a,
+                       const VectorDbl<nDim>& b) {
+  Double sum = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    sum += a(iDim) * b(iDim);
+  }
+  return sum;
+}
+
+template<size_t nDim, class ForwardIterator>
+FORCEINLINE Double dot(ForwardIterator iterator,
+                       const VectorDbl<nDim>& vector) {
+  Double sum = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    sum += *(iterator++) * vector(iDim);
+  }
+  return sum;
+}
+
+template<size_t nDim>
+FORCEINLINE Double squaredNorm(const VectorDbl<nDim>& vector) {
+  Double sum = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    sum += pow(vector(iDim),2);
+  }
+  return sum;
+}
+
+template<size_t nDim, class ForwardIterator>
+FORCEINLINE Double squaredNorm(ForwardIterator iterator) {
+  Double sum = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    sum += pow(*(iterator++),2);
+  }
+  return sum;
+}
+
+template<size_t nDim>
+FORCEINLINE Double norm(const VectorDbl<nDim>& vector) { return sqrt(squaredNorm(vector)); }
+
+template<size_t nDim, class Coord_t>
+FORCEINLINE VectorDbl<nDim> distanceVector(Int iPoint, Int jPoint, const Coord_t& coords) {
+  VectorDbl<nDim> vector_ij;
+  auto coord_i = coords.template innerIter<nDim>(iPoint);
+  auto coord_j = coords.template innerIter<nDim>(jPoint);
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    vector_ij(iDim) = 0.5 * (*(coord_j++) - *(coord_i++));
+  }
+  return vector_ij;
+}
```

Comment:
Those iterators are then used like this, the templating for the known size does require the somewhat baffling syntax you see above...